clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java

```diff
@@ -637,14 +637,55 @@ public void testQuotaMetrics() throws Exception {
     }
 
     @Test
-    public void testGetOffsetsByTimes() {
-        client.prepareResponseFrom(listOffsetResponse(Errors.NONE, 100L, 100L), cluster.leaderFor(tp));
+    public void testGetOffsetsByTimesWithErrors() {
+        // Error code none with unknown offset
+        testGetOffsetsByTimesWithError(Errors.NONE, -1L, 100L, null, 100L);
+        // Error code none with known offset
+        testGetOffsetsByTimesWithError(Errors.NONE, 10L, 100L, 10L, 100L);
+        testGetOffsetsByTimesWithError(Errors.NOT_LEADER_FOR_PARTITION, 10L, 100L, 10L, 100L);
+        testGetOffsetsByTimesWithError(Errors.UNKNOWN_TOPIC_OR_PARTITION, 10L, 100L, 10L, 100L);
+        testGetOffsetsByTimesWithError(Errors.INVALID_REQUEST, 10L, 100L, null, 100L);
+        testGetOffsetsByTimesWithError(Errors.BROKER_NOT_AVAILABLE, 10L, 100L, 10L, 100L);
+    }
+
+    // Test the getOffsetsForTimes behavior when error occurs.
+    // The error is always on partition 0 on the first try.
+    private void testGetOffsetsByTimesWithError(Errors error,
+                                                Long offsetForTp0,
+                                                Long offsetForTp1,
+                                                Long expectedOffsetForTp0,
+                                                Long expectedOffsetForTp1) {
+        client.reset();
+        TopicPartition tp0 = tp;
+        TopicPartition tp1 = new TopicPartition(topicName, 1);
+        // Ensure metadata has both partition.
+        Cluster cluster = TestUtils.clusterWith(2, topicName, 2);
+        metadata.update(cluster, time.milliseconds());
 
-        Map<TopicPartition, OffsetAndTimestamp> offsetAndTimestampMap =
-                fetcher.getOffsetsByTimes(Collections.singletonMap(tp, 0L));
-        assertEquals(offsetAndTimestampMap.get(tp).timestamp(), 100L);
-        assertEquals(offsetAndTimestampMap.get(tp).offset(), 100L);
+        // First try should fail due to metadata error.
+        client.prepareResponseFrom(listOffsetResponse(tp0, error, offsetForTp0, offsetForTp0), cluster.leaderFor(tp0));
+        client.prepareResponseFrom(listOffsetResponse(tp1, Errors.NONE, offsetForTp1, offsetForTp1), cluster.leaderFor(tp1));
+        // Second try should succeed.
+        client.prepareResponseFrom(listOffsetResponse(tp0, Errors.NONE, offsetForTp0, offsetForTp0), cluster.leaderFor(tp0));
+        client.prepareResponseFrom(listOffsetResponse(tp1, Errors.NONE, offsetForTp1, offsetForTp1), cluster.leaderFor(tp1));
+
+        Map<TopicPartition, Long> timestampToSearch = new HashMap<>();
+        timestampToSearch.put(tp0, 0L);
+        timestampToSearch.put(tp1, 0L);
+        Map<TopicPartition, OffsetAndTimestamp> offsetAndTimestampMap = fetcher.getOffsetsByTimes(timestampToSearch);
+        if (expectedOffsetForTp0 == null)
```

Comment:
If we're going to make this function generic, maybe we should have the null check for both partitions?